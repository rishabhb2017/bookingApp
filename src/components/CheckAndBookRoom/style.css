
.btn-section{
    display: flex;
    justify-content: space-between;
}

.btn-submit{
    height:30px; 
    width:80px;
}