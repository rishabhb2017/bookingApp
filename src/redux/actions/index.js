import {
    GET_BOOKINGS,
    ADD_BOOKINGS
} from "./actionTypes";

export const getPosts = () => {
    return {
        type: GET_BOOKINGS,
    };
};
export const addPost = (data) => {
    return {
        type: ADD_BOOKINGS, payload: data
    };
};
