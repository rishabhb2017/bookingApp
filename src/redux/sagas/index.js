export {default as getBookingsSaga} from "./getBookings";
export {default as addBookingsSaga} from "./addBooking";