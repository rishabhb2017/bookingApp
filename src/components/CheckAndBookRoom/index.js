import React, { useCallback, useState, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import ImageIdicator from '../IconIndicator/index';
import './style.css';
import * as actionType from "../../redux/actions/actionTypes";


const CheckAndBookRoom = ({isBooking}) =>{
    
   const availableRoomsArr = [101,102,201,203];    

   const dispatch = useDispatch();

   const bookings = useSelector((state) => state.getBookings.bookings);

   let isBookingConfirmed = useSelector((state) => state.addBookings.bookings);

    const initialState = {
        surname:'',
        room:0,
        dateCal:'',
        isAvaialble:false,
        isVisible:false
    }
    
    const [state, setState]=useState({...initialState});

    useEffect(()=>{
        debugger;
        dispatch({type: actionType.GET_BOOKINGS});
    },[]);
    
    useEffect(()=>{
        if(isBookingConfirmed && +isBookingConfirmed.room === +state.room){
            debugger;
            alert(`Room No. ${isBookingConfirmed.room} Booked Succefully For ${isBookingConfirmed.surname} at ${isBookingConfirmed.dateCal}`);
            setState({...initialState});
            isBookingConfirmed = null;
        }
    },[isBookingConfirmed]);

    function compareDates(dt1, dt2){
     debugger;
     if(dt1.getDate() === dt2.getDate() && dt1.getMonth() === dt2.getMonth() && dt1.getFullYear() === dt2.getFullYear()){
         return true;
     }
      return false
    }

    const onSubmitHandler = (event) =>{
        debugger;
        event.preventDefault();
        const {surname, room, dateCal } = state;
        if(isBooking){
            dispatch({type: actionType.ADD_BOOKINGS, payload:{surname, room: +room, dateCal: new Date(dateCal) }});
        } else {
            if(bookings.find(x=> (compareDates(new Date(x.dateCal), new Date(dateCal)) && x.room === +state.room))){
                setState({...state, isVisible:true, isAvaialble:false});
                alert('Room Not Available');
            }else{
                setState({...state, isVisible:true, isAvaialble:true});
                alert('Room is Available!!');
            }
        }
    };

    const onChangeHandler = (event) =>{
        debugger;
        let name = event.target.name;
        let value = event.target.value;
        setState({...state, [name]: value});
    };

    const surname = isBooking ? <><td>
    <label for="surname">Surname</label>
    </td>
    <td>
    <input aria-label="surname" type="text" id="surname" name="surname" onChange={(e)=>onChangeHandler(e)} value={state.surname} required/>
    </td></> : null;

    return (
        <>
         <form onSubmit={onSubmitHandler}>
             <table>
                 <th>
                   <td><strong><span>{isBooking ? 'ADD BOOKING' : 'CHECK ROOM'}</span></strong></td>  
                 </th>
                 <tr>
                     <td colSpan={2}><hr/></td>
                 </tr>
                 <tr>
                     {surname}
                 </tr>
                 <tr>
                     <td>
                     <label for="room">Room</label>
                     </td>
                     <td>
                     <select id="room" name="room" value={state.room} onChange={onChangeHandler}>
                         <option value={0} disabled selected>Select Room</option>
                     {availableRoomsArr.map((item,indx)=>{
                       return <option value={item} key={indx}>{item}</option>
                     })}
                     </select>
                     </td>
                 </tr>
                 <tr>
                     <td>
                     <label for="dateCal">Date</label>
                     </td>
                     <td>
                     <input name="dateCal" id="dateCal" value={state.date} type="date" required onChange={onChangeHandler} />
                     </td>
                 </tr>
                 <tr>
                     <td></td>
                     <td></td>
                 </tr>
                 <tr>
                     <td></td>
                     <td className={'btn-section'}>
                         <button name={isBooking ? 'Add' : 'Check'} className={'btn-submit'} type='submit'>{isBooking ? 'Add' : 'Check'}</button>
                        <ImageIdicator isVisible={!isBooking && state.isVisible} isAvaialble={state.isAvaialble}/>
                     </td>
                 </tr>
             </table>
         </form>
        </>
    )
}

export default React.memo(CheckAndBookRoom);