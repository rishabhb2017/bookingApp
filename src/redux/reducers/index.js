import { combineReducers } from "redux";

import addBookings from "./addBooking";
import getBookings from "./gotBookings";

export default combineReducers({
    addBookings,
    getBookings
});
