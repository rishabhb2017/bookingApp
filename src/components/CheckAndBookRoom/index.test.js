import React from 'react'
import {render, cleanup, fireEvent, queryByAttribute} from '@testing-library/react'
import CheckAndBookRoom from '../CheckAndBookRoom/index';
import { createStore } from "redux";
import rootReducer from "../../redux/reducers/index";
import { Provider } from 'react-redux';

 afterEach(cleanup)
 
 it('should take a snapshot', () => {

    const CheckAndBookRoomWrapperWithProvider = () => {
        const store = createStore(rootReducer);
        return (
          <Provider store={store}> 
            <CheckAndBookRoom />
          </Provider>
        )
      }
      
    const { asFragment } = render(<CheckAndBookRoomWrapperWithProvider />)
    
    expect(asFragment(<CheckAndBookRoomWrapperWithProvider />)).toMatchSnapshot()
});

it('should surname input exists in Booking Mode only', () => {

    const CheckAndBookRoomWrapperWithProvider = () => {
        const store = createStore(rootReducer);
        return (
          <Provider store={store}> 
            <CheckAndBookRoom isBooking={true} />
          </Provider>
        )
      }
      
    const container = render(<CheckAndBookRoomWrapperWithProvider />);

    const inputEl = container.getByLabelText('surname');

    expect(inputEl.value).toBe('');
});

it('Test surname Text value', () => {

    const CheckAndBookRoomWrapperWithProvider = () => {
        const store = createStore(rootReducer);
        return (
          <Provider store={store}> 
            <CheckAndBookRoom isBooking={true} />
          </Provider>
        )
      }
      
      const container = render(<CheckAndBookRoomWrapperWithProvider />);

      const inputEl = container.getByLabelText('surname');
  
    setTimeout(()=>{
        fireEvent.change(inputEl, {target: {value: 'Bhatnagar'}})
        expect(input.value).toBe('Bhatnagar');
    },200);
   
});
