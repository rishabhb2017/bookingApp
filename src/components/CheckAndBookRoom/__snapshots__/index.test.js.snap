// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should surname text box default be empty 1`] = `
<DocumentFragment>
  <form>
    <table>
      <tbody>
        <tr>
          <th />
          <td>
            <strong>
              <span>
                CHECK ROOM
              </span>
            </strong>
          </td>
        </tr>
        <tr>
          <td
            colspan="2"
          >
            <hr />
          </td>
        </tr>
        <tr />
        <tr>
          <td>
            <label
              for="room"
            >
              Room
            </label>
          </td>
          <td>
            <select
              id="room"
              name="room"
            >
              <option
                disabled=""
                value="0"
              >
                Select Room
              </option>
              <option
                value="101"
              >
                101
              </option>
              <option
                value="102"
              >
                102
              </option>
              <option
                value="201"
              >
                201
              </option>
              <option
                value="203"
              >
                203
              </option>
            </select>
          </td>
        </tr>
        <tr>
          <td>
            <label
              for="dateCal"
            >
              Date
            </label>
          </td>
          <td>
            <input
              id="dateCal"
              name="dateCal"
              required=""
              type="date"
              value=""
            />
          </td>
        </tr>
        <tr>
          <td />
          <td />
        </tr>
        <tr>
          <td />
          <td
            class="btn-section"
          >
            <button
              class="btn-submit"
              name="Check"
              type="submit"
            >
              Check
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </form>
</DocumentFragment>
`;

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <form>
    <table>
      <tbody>
        <tr>
          <th />
          <td>
            <strong>
              <span>
                CHECK ROOM
              </span>
            </strong>
          </td>
        </tr>
        <tr>
          <td
            colspan="2"
          >
            <hr />
          </td>
        </tr>
        <tr />
        <tr>
          <td>
            <label
              for="room"
            >
              Room
            </label>
          </td>
          <td>
            <select
              id="room"
              name="room"
            >
              <option
                disabled=""
                value="0"
              >
                Select Room
              </option>
              <option
                value="101"
              >
                101
              </option>
              <option
                value="102"
              >
                102
              </option>
              <option
                value="201"
              >
                201
              </option>
              <option
                value="203"
              >
                203
              </option>
            </select>
          </td>
        </tr>
        <tr>
          <td>
            <label
              for="dateCal"
            >
              Date
            </label>
          </td>
          <td>
            <input
              id="dateCal"
              name="dateCal"
              required=""
              type="date"
              value=""
            />
          </td>
        </tr>
        <tr>
          <td />
          <td />
        </tr>
        <tr>
          <td />
          <td
            class="btn-section"
          >
            <button
              class="btn-submit"
              name="Check"
              type="submit"
            >
              Check
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </form>
</DocumentFragment>
`;
